Throw an exception when we try to marshal a non-blittable fixed buffer.

jkoritzinsky opened the pull request:
@luqunl and I were talking yesterday and we felt that since we are not supporting correctly marshaling fixed buffers of non-blittable types, we should throw an exception when trying to marshal them instead of having buggy behavior for any non-blittable fixed buffer with more than one element.

Without throwing this exception, we incorrectly lay out the native fields, which can lead to crashes where the native code reads past the (too small) native object and crashes due to an AV or just reads random memory.

For example:

If the native structure is:

```cpp
struct TestStruct {
    BOOL buffer[4];
    INT value;
};
```

and the managed is:

```csharp
struct TestStruct {
    fixed bool buffer[4];
    int value;
}
```

Any native code expecting to receive a `TestStruct` by value will read past the end of the structure if it tries to access the `value` field, possibly causing an AV.


So, we felt that it would be better for the user if we were to throw while marshaling instead of either crashing or causing unexpected behavior (either the issues mentioned above or issues like #20521).

AaronRobinsonMSFT commented on src/vm/fieldmarshaler.h:
```diff
@@ -116,6 +116,8 @@ BOOL HasLayoutMetadata(Assembly* pAssembly, IMDInternalImport *pInternalImport,
 //=======================================================================
 BOOL IsStructMarshalable(TypeHandle th);
 
+BOOL IsFixedBuffer(mdFieldDef field, IMDInternalImport *pInternalImport);
```
Please remove this from the header. This doesn't need to be exported from the compilation unit.

jkoritzinsky commented:
I've added a test onto dotnet/corefx#32538.

jkoritzinsky commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

jkoritzinsky commented:
@dotnet-bot test this please

jkotas commented:
Note that this is a breaking change: It is changing non-throwing behavior into a throwing one. I do not think we should be making breaking changes like this.

jkoritzinsky commented:
That is true. But the behavior when it wasn't throwing was completely incorrect for any fixed buffer with more than one element and would likely cause either crashes or data corruption in the native code by passing effectively junk data that also doesn't have enough space allocated to the native code. I think having an exception early instead of a crash or silent data corruption later is better.

AaronRobinsonMSFT commented:
@jkotas to elaborate on @jkoritzinsky statement and from what I have heard gleaned from conversations with you and others, we are not required to be bug for bug compatible. In this case, the previous behavior was a bug since we weren't marshaling properly when `char` or `bool` were being used. Now instead of marshaling improperly, we are failing fast. Does this mitigate your concern?

jkotas commented:
We still want to maintain high level of compatibility. The following documents describe how we think about it:

https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-changes.md
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#behavioral-changes

No all behavior changes are equal. For example, we consider changing throwing exception into non-throwing exception to be generally fine. The other way around, throwing exception for cases where we did not throw exception before is generally not fine. It does not mean that we never do that. It means that we need to do a bunch more data gathering about the cost/benefit.

For this specific change, consider following program. It will work fine before this change and fails with exception after this change:

```
using System;
using System.Runtime.InteropServices;

unsafe class Test
{   
   struct MyStruct {
       fixed char xxx[10];
   }

   static void Main()
   {
       Console.WriteLine(Marshal.SizeOf(typeof(MyStruct)));
   }
}
```

It is not hard to find places in programs out there that do things like this and that work just fine. E.g. For example, we have them even in our .NET Framework:

https://github.com/Microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/mscorlib/microsoft/win32/win32native.cs#L1335

You can argue that this code is theoretically incorrect, but it does not change anything on the fact that it worked fine for last 10+ years just fine. This change fails on the cost/benefit check for me.

AaronRobinsonMSFT commented:
@jkotas boo... (a) I agree this code is incorrect. (b) your point is well-taken and definitely makes this decision not as clear-cut.

@jkoritzinsky and @luqunl I am inclined to agree with Jan here. Any rebuttals?

jkoritzinsky commented:
I don't have a strong rebuttal, just some observations I found from some quick corner case testing I did based on @jkotas' examples.

Trying to marshal this will throw an exception:
```
   struct MyStruct {
       fixed char xxx[10];
   }
```

However this one (matches the .NET Framework example) won't:

```
  [StructLayout(LayoutKind.Sequential, CharSet= CharSet.Unicode)]
   struct MyStruct {
       fixed char xxx[10];
   }
```


I'm fine reverting this change if we want to. The only rebuttal I have (that we should make it discoverable to users who attempt to marshal these kinds of fixed buffers that they'll almost assuredly get junk data or an AV instead of silently failing and having them file a bug here about it while being very confused) has already been stated above.

AaronRobinsonMSFT commented:
Why does the `CharSet=CharSet.Unicode` example succeed in this case. I think I am missing something subtle here. Does that setting change the blittable nature of the check you added - if it does that seems odd/un-intuitive.

jkoritzinsky commented:
Yep. When the charset is Unicode, `ParseNativeType` assigns the `FieldMarshaler_Copy2` marshaler (one of the always-blittable marshalers) to the field instead of a separate marshaler that does the same thing but technically isn't blittable.
